### PR TITLE
Convert publisher functions to pybind11

### PR DESCRIPTION
**Pull request by cottsay**
Part of #665

Converts:
* rclpy_create_publisher
* rclpy_publisher_get_subscription_count
* rclpy_publisher_get_topic_name
* rclpy_publish
* rclpy_publish_raw

CI (build: --packages-up-to rclpy test: --packages-select rclpy)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13833/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8693/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11540/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13917/)

### CODE REVIEW

**Comment by cottsay** [comment]
I completely missed `rclpy_create_publisher`. Hold off on reviewing this until I get that done.

File: rclpy/src/rclpy/publisher.cpp [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+  rcl_ret_t ret = rcl_publisher_init(
+    &(pub->publisher), node, msg_type,
+    topic.c_str(), &publisher_ops);
+  if (ret != RCL_RET_OK) {
+    if (ret == RCL_RET_TOPIC_NAME_INVALID) {
+      std::string error_text{"Failed to create publisher due to invalid topic name '"};
+      error_text += topic;
+      error_text += "'";
+      throw RCLError(error_text);
```
Comment (sloretz) [inline rclpy/src/rclpy/publisher.cpp:104]: I think it should be `py::value_error` when the topic name is invalid to [match the documentation](https://github.com/ros2/rclpy/pull/695/files#diff-68392ec0fdc86b592da9d13cbf7f93f9ddeafb0b43b5070804574cbdf1b5f851R32).

File: rclpy/src/rclpy/publisher.cpp [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+  }
+  auto pypub = py::reinterpret_steal<py::capsule>(pypub_c);
+  // pypub now owns the rclpy_publisher_t
+  pub.release();
+
+  auto pub_handle = static_cast<rclpy_handle_t *>(pypub);
+  auto node_handle = static_cast<rclpy_handle_t *>(pynode);
+  _rclpy_handle_add_dependency(pub_handle, node_handle);
+  if (PyErr_Occurred()) {
+    _rclpy_handle_dec_ref(node_handle);
```
Comment (sloretz) [inline rclpy/src/rclpy/publisher.cpp:122]: This should be decref'ing `pub_handle` instead of `node_handle`.

**Review by sloretz — CHANGES_REQUESTED** [review]

File: rclpy/src/rclpy/publisher.cpp (hunk 2 above)
Comment (cottsay) [reply rclpy/src/rclpy/publisher.cpp]: 658b6f5

File: rclpy/src/rclpy/publisher.cpp (hunk 1 above)
Comment (cottsay) [reply rclpy/src/rclpy/publisher.cpp]: 658b6f5

**Review by sloretz — APPROVED** [review]

**Pull request merged**